### PR TITLE
Queries for struct complex types

### LINKED ISSUE #31609: Exception projecting value type mapped as a complex type
```C#
var addresses = await context.Orders.Select(e => e.ShippingAddress).ToListAsync();
```

```
Unhandled exception. System.ArgumentException: Argument types do not match
   at System.Linq.Expressions.Expression.Constant(Object value, Type type)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.ProcessEntityShaper(StructuralTypeShaperExpression shaper)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.Inject(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.InjectEntityMaterializers(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.ProcessShaper(Expression shaperExpression, RelationalCommandCache& relationalCommandCache, IReadOnlyList`1& readerColumns, LambdaExpression& relatedDataLoaders, Int32& collectionId)       
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.VisitShapedQuery(ShapedQueryExpression shapedQueryExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Program.<Main>$(String[] args) in C:\local\code\AllTogetherNow\Daily\Daily.cs:line 42
   at Program.<Main>$(String[] args) in C:\local\code\AllTogetherNow\Daily\Daily.cs:line 42
   at Program.<Main>(String[] args)
```

Full repro:

```C#
await using var context = new CustomerContext();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();

var addresses = await context.Orders.Select(e => e.ShippingAddress).ToListAsync();

public class CustomerContext : DbContext
{
    public DbSet<Order> Orders => Set<Order>();

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=AllTogetherNow")
            .EnableSensitiveDataLogging()
            .LogTo(Console.WriteLine, LogLevel.Information);

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Order>(
            b =>
            {
                b.ComplexProperty(e => e.ShippingAddress);
            });
    }
}

public class Order
{
    public int Id { get; set; }
    public Address ShippingAddress { get; set; }
}

public struct Address
{
    public string Line1 { get; init; }
    public string? Line2 { get; init; }
    public string City { get; init; }
    public string Country { get; init; }
    public string PostCode { get; init; }
}
```

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #31609

**Description**

Projecting a complex type represented by a value type throws.

**Customer impact**

Using value types as complex types is a new feature for EF8.

**How found**

Testing.

**Regression**

No.

**Testing**

Tests added for complex types represented by value types.

**Risk**

Small.

### CODE REVIEW

File: src/EFCore.Proxies/Proxies/Internal/ProxyBindingInterceptor.cs [hunk 1]
```diff
@@ -38,7 +38,11 @@ public ProxyBindingInterceptor(IProxyFactory proxyFactory)
     /// </summary>
     public virtual InstantiationBinding ModifyBinding(InstantiationBindingInterceptionData interceptionData, InstantiationBinding binding)
     {
-        var entityType = interceptionData.TypeBase as IEntityType ?? throw new NotImplementedException();
+        if (!(interceptionData.TypeBase is IEntityType entityType))
```
Comment (roji) [inline src/EFCore.Proxies/Proxies/Internal/ProxyBindingInterceptor.cs:41]:
```suggestion
        if (interceptionData.TypeBase is not IEntityType entityType)
```

**Comment by danmoseley** [comment]
thanks, we'll need a template and tactics mail or visit as usual.

**Comment by ajcvickers** [comment]
@danmoseley Are we not on M2 approval for GA?

**Comment by danmoseley** [comment]
No, GA is tactics approval.

**Review by maumar — APPROVED** [review]

**Pull request merged**